File: rust/parity/specs/iac_scanner.py
"""Parity spec for PiIaCScanner.

A spec declares:
  RUST_NAME       - registry key passed to pi_core.run_agent
  run_python(data)-> dict  - run the ORIGINAL python agent, return model_dump()
  SAMPLES         - list of {"input": {...}, "env": {...optional...}}
"""
from _util import load_py_agent

RUST_NAME = "PiIaCScanner"

_mod = load_py_agent("pi_iac_scanner.py")


def run_python(data: dict) -> dict:
    agent = _mod.PiIaCScanner()
    out = agent.scan_iac(_mod.IaCInput(**data))
    return out.model_dump()


# Clean template: no flagged substrings.
_CLEAN = "\n".join(
    [
        'resource "aws_s3_bucket" "b" {',
        '  acl = "private"',
        "  server_side_encryption_configuration {",
        '    sse_algorithm = "aws:kms"',
        "  }",
        "}",
    ]
)

# Public bucket via public-read ACL.
_PUBLIC_READ = 'resource "aws_s3_bucket_acl" "x" {\n  acl = "public-read"\n}'

# Public bucket via wildcard principal (CloudFormation JSON style).
_WILDCARD_PRINCIPAL = '{"Statement": [{"Effect": "Allow", "Principal": "*"}]}'

# Admin ingress: 0.0.0.0/0 together with port 22.
_ADMIN_INGRESS = "\n".join(
    [
        'resource "aws_security_group" "sg" {',
        "  ingress {",
        "    from_port   = 22",
        "    to_port     = 22",
        '    cidr_blocks = ["0.0.0.0/0"]',
        "  }",
        "}",
    ]
)

# Broad network: 0.0.0.0/0 present but NO "22"/"3389"/"cidr_blocks" substrings.
_BROAD_NETWORK = 'allowed_range = "0.0.0.0/0"'

# Disabled encryption.
_NO_ENCRYPTION = 'encryption = "disabled"'

# Combined: public-read + admin ingress + disabled encryption (max risk path).
_COMBINED = "\n".join([_PUBLIC_READ, _ADMIN_INGRESS, _NO_ENCRYPTION])

SAMPLES = [
    {"input": {"file_path": "main.tf", "iac_content": _CLEAN, "iac_type": "terraform"}},
    {"input": {"file_path": "acl.tf", "iac_content": _PUBLIC_READ, "iac_type": "terraform"}},
    {"input": {"file_path": "policy.json", "iac_content": _WILDCARD_PRINCIPAL, "iac_type": "cloudformation"}},
    {"input": {"file_path": "sg.tf", "iac_content": _ADMIN_INGRESS, "iac_type": "terraform"}},
    {"input": {"file_path": "net.tf", "iac_content": _BROAD_NETWORK, "iac_type": "terraform"}},
    {"input": {"file_path": "enc.tf", "iac_content": _NO_ENCRYPTION, "iac_type": "terraform"}},
    {"input": {"file_path": "all.tf", "iac_content": _COMBINED, "iac_type": "terraform"}},
    {"input": {"file_path": "empty.tf", "iac_content": "", "iac_type": "pulumi"}},
    # non-strict env -> high risk but is_secure stays True (WARN path)
    {"input": {"file_path": "acl.tf", "iac_content": _PUBLIC_READ, "iac_type": "terraform"},
     "env": {"PI_IAC_SCANNER_STRICT_MODE": "false"}},
    {"input": {"file_path": "acl.tf", "iac_content": _PUBLIC_READ, "iac_type": "terraform"},
     "env": {"PI_IAC_SCANNER_STRICT_MODE": "true"}},
]

File: rust/crates/pi-agents/src/agents/iac_scanner.rs
//! Port of `pi_micro_agents/pi_iac_scanner.py`.
//!
//! Static analysis of Terraform, CloudFormation, and Pulumi files for exposed
//! ports, public buckets, and missing encryption. Behaviour is a line-for-line
//! mirror of the Python original.

use serde::{Deserialize, Serialize};

#[derive(Debug, Deserialize)]
pub struct Input {
    pub file_path: String,
    pub iac_content: String,
    pub iac_type: String,
}

#[derive(Debug, Serialize, PartialEq)]
pub struct Output {
    pub is_secure: bool,
    pub detected_misconfigs: Vec<String>,
    pub risk_score: f64,
    pub status: String,
}

/// Mirrors `is_strict_mode()`: strict unless the env var is set to a value that
/// is not (case-insensitively) "true".
fn is_strict_mode() -> bool {
    match std::env::var("PI_IAC_SCANNER_STRICT_MODE") {
        Ok(v) => v.to_lowercase() == "true",
        Err(_) => true,
    }
}

pub fn scan_iac(input: &Input) -> Output {
    let content = &input.iac_content;
    let mut misconfigs: Vec<String> = Vec::new();
    let mut risk_score: f64 = 0.0;

    // Public buckets/ACL check
    if content.contains("public-read")
        || content.contains("Principal\": \"*\"")
        || content.contains("Principal\":\"*\"")
        || content.contains("Principal = \"*\"")
    {
        misconfigs.push(
            "Public Access: S3/Blob storage resource configured with public access or wildcard principal."
                .to_string(),
        );
        risk_score = risk_score.max(85.0);
    }

    // Overly broad ingress ports (e.g. port 22 or 3389 open to 0.0.0.0/0)
    if content.contains("0.0.0.0/0") {
        if content.contains("22") || content.contains("3389") || content.contains("cidr_blocks") {
            misconfigs.push(
                "Exposed Ingress: Administrative ports (22/3389) open to wildcard range (0.0.0.0/0)."
                    .to_string(),
            );
            risk_score = risk_score.max(90.0);
        } else {
            misconfigs.push("Broad Network: Generic wildcard network ingress allowed.".to_string());
            risk_score = risk_score.max(40.0);
        }
    }

    // Missing or disabled encryption
    if content.contains("encryption = \"disabled\"")
        || content.contains("sse_algorithm = \"none\"")
        || content.contains("encryption\": \"false\"")
    {
        misconfigs.push(
            "Unencrypted Resource: Data-at-rest encryption is explicitly disabled or not configured."
                .to_string(),
        );
        risk_score = risk_score.max(75.0);
    }

    let mut is_sec = true;
    if risk_score > 30.0 && is_strict_mode() {
        is_sec = false;
    }

    let mut status = if is_sec {
        "PASSED".to_string()
    } else {
        "FAILED_COMPLIANCE".to_string()
    };
    if risk_score > 0.0 && is_sec {
        status = "WARN_COMPLIANCE".to_string();
    }

    Output {
        is_secure: is_sec,
        detected_misconfigs: misconfigs,
        risk_score,
        status,
    }
}

pub fn run_json(input_json: &str) -> Result<String, String> {
    let input: Input = serde_json::from_str(input_json).map_err(|e| e.to_string())?;
    let out = scan_iac(&input);
    serde_json::to_string(&out).map_err(|e| e.to_string())
}

#[cfg(test)]
mod tests {
    use super::*;

    fn run(content: &str) -> Output {
        scan_iac(&Input {
            file_path: "main.tf".into(),
            iac_content: content.into(),
            iac_type: "terraform".into(),
        })
    }

    #[test]
    fn clean_template_passes() {
        let o = run("resource \"aws_s3_bucket\" \"b\" {\n  acl = \"private\"\n}");
        assert!(o.is_secure);
        assert_eq!(o.status, "PASSED");
        assert_eq!(o.risk_score, 0.0);
        assert!(o.detected_misconfigs.is_empty());
    }

    #[test]
    fn public_bucket_flagged() {
        let o = run("acl = \"public-read\"");
        assert!(!o.is_secure);
        assert_eq!(o.status, "FAILED_COMPLIANCE");
        assert_eq!(o.risk_score, 85.0);
        assert_eq!(o.detected_misconfigs.len(), 1);
    }

    #[test]
    fn admin_ingress_flagged() {
        let o = run("cidr_blocks = [\"0.0.0.0/0\"]\nfrom_port = 22");
        assert!(!o.is_secure);
        assert_eq!(o.risk_score, 90.0);
    }

    #[test]
    fn broad_network_only() {
        // 0.0.0.0/0 present but no "22"/"3389"/"cidr_blocks" substrings.
        // risk 40.0 > 30.0, and strict mode defaults to true, so is_secure -> false.
        let o = run("network = \"0.0.0.0/0\"");
        assert_eq!(o.risk_score, 40.0);
        assert!(!o.is_secure);
        assert_eq!(o.status, "FAILED_COMPLIANCE");
        assert_eq!(o.detected_misconfigs.len(), 1);
    }
}
